src/main/groovy/io/openliberty/tools/gradle/tasks/AbstractServerTask.groovy

```diff
@@ -731,21 +745,103 @@ abstract class AbstractServerTask extends AbstractTask {
         }
     }
 
-    private void writeServerEnvProperties(File file, Properties properties, Properties projectProperties) throws IOException {
-        Properties combinedEnvProperties = null
+    private void handleServerEnvFileAndProperties(String serverEnvPath, String serverDirectory) {
+        File envFile = new File(serverDirectory, "server.env")
+        Properties configuredProps = combineServerEnvProperties(server.env, envProjectProps);
+
+        if(server.appendServerEnv) {
+            setServerEnvWithAppendServerEnvHelper(envFile, serverEnvPath, configuredProps)
+        }
+        else {
+            setServerEnvHelper(envFile, serverEnvPath, configuredProps)
+        }
+    }
+
+    private void setServerEnvWithAppendServerEnvHelper(File envFile, String serverEnvPath, Properties configuredProps) {
+        Properties serverEnvProps = convertServerEnvToProperties(envFile);
+        Properties mergedProperties = new Properties();
+
+        if (server.serverEnvFile != null && server.serverEnvFile.exists()) {
+            if (serverEnvPath != null) {
+                logger.warn("The " + serverEnvPath + " file is merged with the " + server.serverEnvFile.getCanonicalPath() + " file.")
+            }
+            Properties configuredServerEnvProps = convertServerEnvToProperties(server.serverEnvFile);
+            //Merge with either default server.env or with what has already been merged if
+            mergedProperties = (Properties) combineServerEnvProperties(serverEnvProps, configuredServerEnvProps);
+        }
+
+        if ((server.env != null && !server.env.isEmpty()) || !envProjectProps.isEmpty()) {
+            if (serverEnvPath != null) {
+                logger.warn("The " + serverEnvPath + " file is merged with inlined configuration.")
+            }
+
+            mergedProperties = combineServerEnvProperties(mergedProperties, configuredProps);
+        }
+
+        if(!mergedProperties.isEmpty()) {
+            writeServerEnvProperties(envFile, mergedProperties);
+        }
+    }
+
+    private void setServerEnvHelper(File envFile, String serverEnvPath, Properties configuredProps) {
+        if ((server.env != null && !server.env.isEmpty()) || !envProjectProps.isEmpty()) {
+            if (serverEnvPath != null) {
+                logger.warn("The " + serverEnvPath + " file is overwritten by inlined configuration.")
+            }
+            writeServerEnvProperties(envFile, configuredProps)
+            serverEnvPath = "inlined configuration"
```

Comment:
Strings are immutable. Setting this has no affect on the serverEnvPath variable that was passed in. Perhaps we need to be passing this back as the return value?

Reply:
Will do. Especially since I'm modifying `setServerEnvWithAppendServerEnvHelper` to also return this value.